mlir/lib/Dialect/MQTOpt/Transforms/ToQuantumComputationPattern.cpp

```diff
@@ -321,26 +374,28 @@ struct ToQuantumComputationPattern final : mlir::OpRewritePattern<AllocOp> {
 
     // Update the inputs of all non-mqtopt operations that use mqtopt operations
     // as inputs, as these will be deleted later.
-    // NOLINTNEXTLINE(bugprone-nondeterministic-pointer-iteration-order)
-    for (auto* operation : visited) {
-      if (operation->getDialect()->getNamespace() != DIALECT_NAME_MQTOPT) {
-        updateMQTOptInputs(*operation, rewriter, newAlloc.getQureg(),
-                           measureCount);
-      }
+    for (auto* operation : llvm::make_early_inc_range(mqtUsers)) {
+      mqtUsers.erase(operation); // safe deletion as op will be erased next
+      updateMQTOptInputs(*operation, rewriter, newAlloc.getQureg());
     }
 
     // Delete all operations that are part of the mqtopt dialect (except for
     // `AllocOp`).
-    // NOLINTNEXTLINE(bugprone-nondeterministic-pointer-iteration-order)
-    for (auto* operation : visited) {
-      if (operation->getDialect()->getNamespace() == DIALECT_NAME_MQTOPT) {
-        deleteRecursively(*operation, rewriter);
+    visited.erase(op); // erase alloc op
+    while (!visited.empty()) {
+      // Enable updates of `visited` set in the loop.
+      for (auto* operation : llvm::make_early_inc_range(visited)) {
+        if (operation->getDialect()->getNamespace() == DIALECT_NAME_MQTOPT) {
+          if (operation->getUsers().empty()) {
+            visited.erase(operation);
+            rewriter.eraseOp(
+                operation); // deletes all definitions of defining OPs
+          }
+        }
```

Comment:
This seems like a confusing way to write this removal here.
The while loop has no break conditions, so the `visited` container` must eventually be empty for this function to return.
Yet, the inner loop has two special if conditions that check for some kind of property before deciding to remove an operation.
Maybe I am overlooking things, but isn't there a clearer way to write this?